### PR TITLE
delete --disable_loss_parallel flag

### PR DESCRIPTION
**Pull request by pianpwk**
loss migration: loss-parallel custom autograd handling, DTensor local_map / unwrapping, now lives under cross_entropy_loss fn

loss-parallel flag deletion: 
- for trainers, TP on = LP on, uses custom autograd impl instead of pytorch's implicit `loss_parallel()` context. Sets model sharding configs to return vocab-sharded logits: `tp_gather_logits=False`
- RL usecases require full logits from lm_head, `tp_gather_logits=True`
- GraphTrainer TP on = LP on

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #3690
* #3673
* #3655
* #3664
* #3654
* #3278
* #3472
* __->__ #3694

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -510,11 +515,8 @@ def __init__(self, config: Config):
             base_folder=config.dump_folder,
         )
 
-        loss_parallel_enabled = (
-            parallel_dims.tp_enabled and not config.parallelism.disable_loss_parallel
-        )
         self.train_context = dist_utils.get_train_context(
-            enable_loss_parallel=loss_parallel_enabled,
+            enable_loss_parallel=parallel_dims.tp_enabled,
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: remove this flag `enable_loss_parallel` globally -- it is always enabled

Comment (pianpwk) [reply torchtitan/trainer.py]: uhhh reshuffling things a bit while I figure this out but...
- ended up keeping the top-level `disable_loss_parallel` flag
- require LP for TP trainer (torchft too)
- delete flag from dist_utils.train_context
- renamed sharding config flag to `gather_tp_logits` - appears both RL policy trainer & generator expect full logits from lm_head, so `is_inference` is misleading
- main blocker is graph trainer - unclear what they want, though no reliance on train_context

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -35,6 +36,20 @@ def cross_entropy_loss(
...
+        assert get_spmd_backend() == "default"
+        if loss_parallel:
+            assert pred.placements == (Shard(pred.ndim - 1),)
+            return _LossParallelCrossEntropy.apply(
+                pred.to_local().flatten(0, 1).float(),
+                labels.flatten(0, 1),
+                pred.device_mesh.get_group("tp"),
+                pred.shape[-1],
+            )
+        else:
```
Comment (tianyu-l) [inline torchtitan/components/loss.py]: when would we hit this? We should delete loss_parallel flag.

File: torchtitan/experiments/graph_trainer/precompile_main.py [hunk 3]
```diff
@@ -233,18 +232,9 @@ def _precompile_aot_fx_trace(
...
-    # Enable loss_parallel when TP is active and loss_parallel is not
-    # disabled. This matches the training path which wraps tracing +
-    # execution inside train_context() → loss_parallel(). Without it,
-    # cross_entropy fails with "mixed torch.Tensor and DTensor" because
-    # the TP-parallelized model outputs Shard'd DTensors but labels
-    # remain plain tensors.
-    loss_parallel_enabled = (
-        parallel_dims.tp_enabled and not config.parallelism.disable_loss_parallel
-    )
     loss_parallel_ctx = (
```
Comment (tianyu-l) [inline torchtitan/experiments/graph_trainer/precompile_main.py:251]: This ctx should be removed since we migrate to megatron-style loss parallel? Can leave a TODO. cc @bobrenjc93

File: torchtitan/experiments/rl/actors/trainer.py [hunk 4]
```diff
@@ -313,7 +313,7 @@ def _build_model(
         # Fill sharding configs on the config BEFORE build via the
         # model-agnostic `update_from_config` hook (RL's trainer bypasses
         # `torchtitan.Trainer's` call, so we invoke it directly).
-        model_spec.model.update_from_config(config=config)
+        model_spec.model.update_from_config(config=config, tp_gather_logits=True)
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/trainer.py]: instead of doing this, let's change this code to TP dim all-gather 
https://github.com/pytorch/torchtitan/blob/b70844d5f74e538800fd62c35eb6699d211e05e7/torchtitan/experiments/rl/actors/trainer.py#L61

File: torchtitan/experiments/rl/models/vllm_wrapper.py [hunk 5]
```diff
@@ -191,7 +191,10 @@ def __init__(
         class _InferenceConfig:
             parallelism: ParallelismConfig
 
-        self.config.update_from_config(config=_InferenceConfig(parallelism=parallelism))
+        self.config.update_from_config(
+            config=_InferenceConfig(parallelism=parallelism),
+            tp_gather_logits=True,
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/models/vllm_wrapper.py]: Instead of doing this, let's update the `logits = logits.to_local()` below to TP dim all-gather.

File: torchtitan/models/common/decoder_sharding.py [hunk 6]
```diff
@@ -277,13 +277,14 @@ def set_decoder_sharding_config(
...
     unsharded (equivalent to the legacy ``NoParallel`` plan).
+    ``tp_gather_logits=True`` -> ``lm_head`` returns full logits.
     """
     activation_layout = (
         dense_sequence_parallel_placement()
         if enable_sp
         else dense_activation_placement(tp=spmd.I)
     )
-    loss_tp = spmd.S(-1) if loss_parallel else spmd.I
+    loss_tp = spmd.I if tp_gather_logits else spmd.S(-1)
```
Comment (tianyu-l) [inline torchtitan/models/common/decoder_sharding.py]: Let's not introduce `tp_gather_logits`. Instead, let's always do S(-1) and give the all-gather responsibility to RL trainer / generator.
- This matches our vocab parallel embedding behavior
- Even in RL trainer people may want to have complicated chunked loss https://github.com/pytorch/torchtitan/pull/3477 (+ potentially loss paralle?)

I feel the scope of such all-gather is pretty limited (right after RL trainer / generator exits), so it's not worth spreading this config everywhere. Hope that makes sense.

**Review by tianyu-l — APPROVED** [review]
Please remove `tp_gather_logits` and always do S(-1).

Please let me know if you disagree.
